[PAC][lld] Do not emit warnings for `-z pac-plt` with valid PAuth core info

kovdan01 opened the pull request:
When PAuth core info is present and (platform,version) is not (0,0),
treat input files as pac-enabled and do not emit a warning with
`-z pac-plt` passed.

kovdan01 committed:
[PAC][lld] Do not emit warnings for `-z pac-plt` with valid PAuth core info

When PAuth core info is present and (platform,version) is not (0,0),
treat input files as pac-enabled and do not emit a warning with
`-z pac-plt` passed.

kovdan01 commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112959?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#112959** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112959?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#112958** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112958?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @kovdan01 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

DanielKristofKiss commented on lld/test/ELF/aarch64-feature-pauth.s:
```diff
@@ -33,13 +33,53 @@
+# RUN: ld.lld tag-zero.o func2.o func3.so -z pac-plt --shared -o pacplt-warn 2>&1 | FileCheck --check-prefix WARN2 %s
+
+# WARN2:      warning: tag-zero.o: -z pac-plt: file does not have GNU_PROPERTY_AARCH64_FEATURE_1_PAC property and no valid PAuth core info present for this link job
+# WARN2-NEXT: warning: func2.o: -z pac-plt: file does not have GNU_PROPERTY_AARCH64_FEATURE_1_PAC property and no valid PAuth core info present for this link job
```
why this warning is not present for linking `pacplt-nowarn` with --fatal-warnings as func2.o is the same for both link command?

kovdan01 replied:
@DanielKristofKiss We look at the common PAuth core info of all the input files, not the PAuth core info of individual files to determine if a warning is needed. The common info is calculated as follows: if there are two files with different info, emit an error, otherwise, if the info is the same for all the files where it's present, the common core info is set to that value. For files with no PAuth core info, we emit nothing/warning/error depending on `-z pauth-report` option - see #72714

The motivation for using common PAuth core info instead of individual ones is that it allows to have assembly files like crti.s and crtn.s unmodified. This way, we do not need to manually define GNU property with PAuth core info in these files to avoid warnings.

kovdan01 committed:
Omit unneeded `!ctx.aarch64PauthAbiCoreInfo.empty()` check

With `llvm::any_of(ctx.aarch64PauthAbiCoreInfo, /*...*/)` being true,
the byte range is always non-empty.

DanielKristofKiss reviewed (approved):
lgtm

kovdan01 commented:
### Merge activity

* **Nov 10, 1:05 AM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/112959).

MaskRay commented on lld/test/ELF/aarch64-feature-pauth.s:
```diff
@@ -33,13 +33,53 @@
+# RUN: llvm-mc -filetype=obj -triple=aarch64-linux-gnu %p/Inputs/aarch64-func2.s -o func2.o
+# RUN: llvm-mc -filetype=obj -triple=aarch64-linux-gnu %p/Inputs/aarch64-func3.s -o func3.o
+# RUN: ld.lld func3.o --shared -o func3.so
+# RUN: ld.lld tag1.o     func2.o func3.so -z pac-plt --shared -o pacplt-nowarn --fatal-warnings
```
excess spaces

kovdan01 replied:
Removed in 58b0ffa. Initially added to align with the next RUN line, but I'm OK with deleting the spaces if it's considered not needed.

kovdan01 committed:
Remove excess spaces

kovdan01 commented:
Both buildbot failures present at the moment look not relates to this PR and are probably caused by some CI instabilities:

1. https://lab.llvm.org/buildbot/#/builders/4/builds/3472 - time out on running test-suite
2. https://lab.llvm.org/buildbot/#/builders/55/builds/3462 - on hwasan build, line `==153460==Unable to get registers from thread 153361` appearing in stderr cause a lit test failure.
